{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BCRs input file, BCR_original_emb file, and rna input file\n",
    "1. Ensure that each file includes an index column labeled \"barcode\". This column serves as a unique identifier for each cell.\n",
    "2. Verify that the cells are aligned in the same order across all three files.\n",
    "3. The BCR file must include the following columns: \"fwr1\", \"cdr1\", \"cdr2\", \"fwr2\", \"cdr3\", \"fwr3\", and \"fwr4\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcr = pd.read_csv(\"exampledata/example_bcr.csv\", index_col=\"barcode\")\n",
    "rna = pd.read_csv(\"exampledata/example_rna.csv\", index_col=\"barcode\")\n",
    "assert(bcr.index.tolist() == rna.index.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>...</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "      <th>exact_subclonotype_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>label</th>\n",
       "      <th>whole_seq</th>\n",
       "      <th>new_whole_seq</th>\n",
       "      <th>v_identity</th>\n",
       "      <th>SHM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAATAGCAA-1_06</th>\n",
       "      <td>AAACCTGCAATAGCAA-1_contig_1_06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>619</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV2-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>clonotype455</td>\n",
       "      <td>clonotype455_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>spikepositive</td>\n",
       "      <td>QITLKESGPTLVKPTQTLTLTCTFSGFSLTSGMGVGWIRQPPGKAL...</td>\n",
       "      <td>06:QITLKESGPTLVKPTQTLTLTCTFSGFSLTSGMGVGWIRQPPG...</td>\n",
       "      <td>0.03754</td>\n",
       "      <td>0.061453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACAACTGT-1_06</th>\n",
       "      <td>AAACCTGCACAACTGT-1_contig_2_06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>537</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV1-24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>52</td>\n",
       "      <td>clonotype278</td>\n",
       "      <td>clonotype278_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>spikepositive</td>\n",
       "      <td>QVQLVQSGAEVKKPGASVKVSCKVSGYTLTELSMHWVRQAPGKGLE...</td>\n",
       "      <td>06:QVQLVQSGAEVKKPGASVKVSCKVSGYTLTELSMHWVRQAPGK...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.008596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCCTGTG-1_06</th>\n",
       "      <td>AAACCTGCAGCCTGTG-1_contig_1_06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>656</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV4-59</td>\n",
       "      <td>IGHD3-22</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>clonotype1394</td>\n",
       "      <td>clonotype1394_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>spikepositive</td>\n",
       "      <td>QVQLQESGPGLVKPSETLSLICTVSGGSISSYYWSWIRQPAGKGLE...</td>\n",
       "      <td>06:QVQLQESGPGLVKPSETLSLICTVSGGSISSYYWSWIRQPAGK...</td>\n",
       "      <td>0.02397</td>\n",
       "      <td>0.066489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGTCAGCC-1_06</th>\n",
       "      <td>AAACCTGCAGTCAGCC-1_contig_2_06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>670</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV1-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>clonotype969</td>\n",
       "      <td>clonotype969_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>spikepositive</td>\n",
       "      <td>QVQLVQSGAEVKNIGVSVKVSCKASGYTFTDYYIHWVRQAPGQGLE...</td>\n",
       "      <td>06:QVQLVQSGAEVKNIGVSVKVSCKASGYTFTDYYIHWVRQAPGQ...</td>\n",
       "      <td>0.06081</td>\n",
       "      <td>0.095368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTCCTCCA-1_06</th>\n",
       "      <td>AAACCTGGTTCCTCCA-1_contig_2_06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>692</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV1-18</td>\n",
       "      <td>IGHD3-22</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>clonotype1466</td>\n",
       "      <td>clonotype1466_consensus_1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>spikepositive</td>\n",
       "      <td>QVQLVQSGDEVKKPGASVKVSCEASGYTFISYGIAWVRQAPGQGLE...</td>\n",
       "      <td>06:QVQLVQSGDEVKKPGASVKVSCEASGYTFISYGIAWVRQAPGQ...</td>\n",
       "      <td>0.11576</td>\n",
       "      <td>0.139594</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            contig_id  is_cell  \\\n",
       "barcode                                                          \n",
       "AAACCTGCAATAGCAA-1_06  AAACCTGCAATAGCAA-1_contig_1_06     True   \n",
       "AAACCTGCACAACTGT-1_06  AAACCTGCACAACTGT-1_contig_2_06     True   \n",
       "AAACCTGCAGCCTGTG-1_06  AAACCTGCAGCCTGTG-1_contig_1_06     True   \n",
       "AAACCTGCAGTCAGCC-1_06  AAACCTGCAGTCAGCC-1_contig_2_06     True   \n",
       "AAACCTGGTTCCTCCA-1_06  AAACCTGGTTCCTCCA-1_contig_2_06     True   \n",
       "\n",
       "                       high_confidence  length chain    v_gene    d_gene  \\\n",
       "barcode                                                                    \n",
       "AAACCTGCAATAGCAA-1_06             True     619   IGH   IGHV2-5       NaN   \n",
       "AAACCTGCACAACTGT-1_06             True     537   IGH  IGHV1-24       NaN   \n",
       "AAACCTGCAGCCTGTG-1_06             True     656   IGH  IGHV4-59  IGHD3-22   \n",
       "AAACCTGCAGTCAGCC-1_06             True     670   IGH   IGHV1-2       NaN   \n",
       "AAACCTGGTTCCTCCA-1_06             True     692   IGH  IGHV1-18  IGHD3-22   \n",
       "\n",
       "                      j_gene c_gene  full_length  ...  umis raw_clonotype_id  \\\n",
       "barcode                                           ...                          \n",
       "AAACCTGCAATAGCAA-1_06  IGHJ4  IGHG1         True  ...     5     clonotype455   \n",
       "AAACCTGCACAACTGT-1_06  IGHJ4   IGHM         True  ...    52     clonotype278   \n",
       "AAACCTGCAGCCTGTG-1_06  IGHJ4  IGHG1         True  ...     8    clonotype1394   \n",
       "AAACCTGCAGTCAGCC-1_06  IGHJ5  IGHG1         True  ...     8     clonotype969   \n",
       "AAACCTGGTTCCTCCA-1_06  IGHJ4  IGHG1         True  ...    21    clonotype1466   \n",
       "\n",
       "                                raw_consensus_id exact_subclonotype_id sample  \\\n",
       "barcode                                                                         \n",
       "AAACCTGCAATAGCAA-1_06   clonotype455_consensus_1                     1      6   \n",
       "AAACCTGCACAACTGT-1_06   clonotype278_consensus_1                     1      6   \n",
       "AAACCTGCAGCCTGTG-1_06  clonotype1394_consensus_1                     1      6   \n",
       "AAACCTGCAGTCAGCC-1_06   clonotype969_consensus_1                     1      6   \n",
       "AAACCTGGTTCCTCCA-1_06  clonotype1466_consensus_1                     1      6   \n",
       "\n",
       "                               label  \\\n",
       "barcode                                \n",
       "AAACCTGCAATAGCAA-1_06  spikepositive   \n",
       "AAACCTGCACAACTGT-1_06  spikepositive   \n",
       "AAACCTGCAGCCTGTG-1_06  spikepositive   \n",
       "AAACCTGCAGTCAGCC-1_06  spikepositive   \n",
       "AAACCTGGTTCCTCCA-1_06  spikepositive   \n",
       "\n",
       "                                                               whole_seq  \\\n",
       "barcode                                                                    \n",
       "AAACCTGCAATAGCAA-1_06  QITLKESGPTLVKPTQTLTLTCTFSGFSLTSGMGVGWIRQPPGKAL...   \n",
       "AAACCTGCACAACTGT-1_06  QVQLVQSGAEVKKPGASVKVSCKVSGYTLTELSMHWVRQAPGKGLE...   \n",
       "AAACCTGCAGCCTGTG-1_06  QVQLQESGPGLVKPSETLSLICTVSGGSISSYYWSWIRQPAGKGLE...   \n",
       "AAACCTGCAGTCAGCC-1_06  QVQLVQSGAEVKNIGVSVKVSCKASGYTFTDYYIHWVRQAPGQGLE...   \n",
       "AAACCTGGTTCCTCCA-1_06  QVQLVQSGDEVKKPGASVKVSCEASGYTFISYGIAWVRQAPGQGLE...   \n",
       "\n",
       "                                                           new_whole_seq  \\\n",
       "barcode                                                                    \n",
       "AAACCTGCAATAGCAA-1_06  06:QITLKESGPTLVKPTQTLTLTCTFSGFSLTSGMGVGWIRQPPG...   \n",
       "AAACCTGCACAACTGT-1_06  06:QVQLVQSGAEVKKPGASVKVSCKVSGYTLTELSMHWVRQAPGK...   \n",
       "AAACCTGCAGCCTGTG-1_06  06:QVQLQESGPGLVKPSETLSLICTVSGGSISSYYWSWIRQPAGK...   \n",
       "AAACCTGCAGTCAGCC-1_06  06:QVQLVQSGAEVKNIGVSVKVSCKASGYTFTDYYIHWVRQAPGQ...   \n",
       "AAACCTGGTTCCTCCA-1_06  06:QVQLVQSGDEVKKPGASVKVSCEASGYTFISYGIAWVRQAPGQ...   \n",
       "\n",
       "                      v_identity       SHM  \n",
       "barcode                                     \n",
       "AAACCTGCAATAGCAA-1_06    0.03754  0.061453  \n",
       "AAACCTGCACAACTGT-1_06    0.00000  0.008596  \n",
       "AAACCTGCAGCCTGTG-1_06    0.02397  0.066489  \n",
       "AAACCTGCAGTCAGCC-1_06    0.06081  0.095368  \n",
       "AAACCTGGTTCCTCCA-1_06    0.11576  0.139594  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IGJ</th>\n",
       "      <th>IGLL5</th>\n",
       "      <th>RP11.685N3.1</th>\n",
       "      <th>RP11.731F5.2</th>\n",
       "      <th>CH17.132F21.1</th>\n",
       "      <th>HIST1H4C</th>\n",
       "      <th>UBE2C</th>\n",
       "      <th>DERL3</th>\n",
       "      <th>TUBA1B</th>\n",
       "      <th>MKI67</th>\n",
       "      <th>...</th>\n",
       "      <th>AZI2</th>\n",
       "      <th>HHLA3</th>\n",
       "      <th>BFSP2.AS1</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>TARSL2</th>\n",
       "      <th>DIS3</th>\n",
       "      <th>CTD.2256P15.2</th>\n",
       "      <th>ZNF283</th>\n",
       "      <th>FAHD2B</th>\n",
       "      <th>RP11.485G4.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAATAGCAA-1_06</th>\n",
       "      <td>0.345539</td>\n",
       "      <td>0.411600</td>\n",
       "      <td>-0.086132</td>\n",
       "      <td>-0.151236</td>\n",
       "      <td>0.034590</td>\n",
       "      <td>0.737681</td>\n",
       "      <td>0.000468</td>\n",
       "      <td>0.122183</td>\n",
       "      <td>1.349319</td>\n",
       "      <td>0.002903</td>\n",
       "      <td>...</td>\n",
       "      <td>0.771746</td>\n",
       "      <td>-0.000238</td>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.016492</td>\n",
       "      <td>0.660869</td>\n",
       "      <td>-0.059796</td>\n",
       "      <td>0.012625</td>\n",
       "      <td>0.007151</td>\n",
       "      <td>0.014972</td>\n",
       "      <td>0.026571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACAACTGT-1_06</th>\n",
       "      <td>0.117444</td>\n",
       "      <td>0.985152</td>\n",
       "      <td>-0.210269</td>\n",
       "      <td>-0.231358</td>\n",
       "      <td>-0.074831</td>\n",
       "      <td>1.304759</td>\n",
       "      <td>-0.003087</td>\n",
       "      <td>0.564784</td>\n",
       "      <td>0.796385</td>\n",
       "      <td>0.000614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.126555</td>\n",
       "      <td>-0.057113</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.127062</td>\n",
       "      <td>0.248703</td>\n",
       "      <td>0.093876</td>\n",
       "      <td>0.005728</td>\n",
       "      <td>0.025548</td>\n",
       "      <td>0.009291</td>\n",
       "      <td>-0.002509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCCTGTG-1_06</th>\n",
       "      <td>0.118185</td>\n",
       "      <td>0.304202</td>\n",
       "      <td>1.553921</td>\n",
       "      <td>-0.071504</td>\n",
       "      <td>-0.152440</td>\n",
       "      <td>1.007265</td>\n",
       "      <td>-0.005297</td>\n",
       "      <td>0.122008</td>\n",
       "      <td>1.350721</td>\n",
       "      <td>0.001795</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008371</td>\n",
       "      <td>-0.001826</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>-0.070642</td>\n",
       "      <td>-0.085906</td>\n",
       "      <td>0.005892</td>\n",
       "      <td>-0.018601</td>\n",
       "      <td>0.030213</td>\n",
       "      <td>-0.042815</td>\n",
       "      <td>-0.001975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGTCAGCC-1_06</th>\n",
       "      <td>0.123230</td>\n",
       "      <td>1.773176</td>\n",
       "      <td>-0.150341</td>\n",
       "      <td>-0.035437</td>\n",
       "      <td>0.003177</td>\n",
       "      <td>0.666422</td>\n",
       "      <td>0.046367</td>\n",
       "      <td>0.030498</td>\n",
       "      <td>1.141658</td>\n",
       "      <td>0.000546</td>\n",
       "      <td>...</td>\n",
       "      <td>0.069885</td>\n",
       "      <td>-0.079806</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.015928</td>\n",
       "      <td>0.099969</td>\n",
       "      <td>0.084648</td>\n",
       "      <td>0.014874</td>\n",
       "      <td>-0.017749</td>\n",
       "      <td>0.041447</td>\n",
       "      <td>0.007477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTCCTCCA-1_06</th>\n",
       "      <td>0.162600</td>\n",
       "      <td>0.300888</td>\n",
       "      <td>-0.164492</td>\n",
       "      <td>0.755047</td>\n",
       "      <td>-0.039643</td>\n",
       "      <td>0.022945</td>\n",
       "      <td>0.017033</td>\n",
       "      <td>0.154383</td>\n",
       "      <td>0.842697</td>\n",
       "      <td>0.007529</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025873</td>\n",
       "      <td>0.010463</td>\n",
       "      <td>-0.009411</td>\n",
       "      <td>0.573100</td>\n",
       "      <td>-0.028001</td>\n",
       "      <td>0.097279</td>\n",
       "      <td>0.011172</td>\n",
       "      <td>0.013083</td>\n",
       "      <td>0.161455</td>\n",
       "      <td>-0.006934</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 5000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            IGJ     IGLL5  RP11.685N3.1  RP11.731F5.2  \\\n",
       "barcode                                                                 \n",
       "AAACCTGCAATAGCAA-1_06  0.345539  0.411600     -0.086132     -0.151236   \n",
       "AAACCTGCACAACTGT-1_06  0.117444  0.985152     -0.210269     -0.231358   \n",
       "AAACCTGCAGCCTGTG-1_06  0.118185  0.304202      1.553921     -0.071504   \n",
       "AAACCTGCAGTCAGCC-1_06  0.123230  1.773176     -0.150341     -0.035437   \n",
       "AAACCTGGTTCCTCCA-1_06  0.162600  0.300888     -0.164492      0.755047   \n",
       "\n",
       "                       CH17.132F21.1  HIST1H4C     UBE2C     DERL3    TUBA1B  \\\n",
       "barcode                                                                        \n",
       "AAACCTGCAATAGCAA-1_06       0.034590  0.737681  0.000468  0.122183  1.349319   \n",
       "AAACCTGCACAACTGT-1_06      -0.074831  1.304759 -0.003087  0.564784  0.796385   \n",
       "AAACCTGCAGCCTGTG-1_06      -0.152440  1.007265 -0.005297  0.122008  1.350721   \n",
       "AAACCTGCAGTCAGCC-1_06       0.003177  0.666422  0.046367  0.030498  1.141658   \n",
       "AAACCTGGTTCCTCCA-1_06      -0.039643  0.022945  0.017033  0.154383  0.842697   \n",
       "\n",
       "                          MKI67  ...      AZI2     HHLA3  BFSP2.AS1    PIK3CA  \\\n",
       "barcode                          ...                                            \n",
       "AAACCTGCAATAGCAA-1_06  0.002903  ...  0.771746 -0.000238   0.000177  0.016492   \n",
       "AAACCTGCACAACTGT-1_06  0.000614  ...  0.126555 -0.057113   0.000036  0.127062   \n",
       "AAACCTGCAGCCTGTG-1_06  0.001795  ...  0.008371 -0.001826   0.000120 -0.070642   \n",
       "AAACCTGCAGTCAGCC-1_06  0.000546  ...  0.069885 -0.079806   0.000041  0.015928   \n",
       "AAACCTGGTTCCTCCA-1_06  0.007529  ...  0.025873  0.010463  -0.009411  0.573100   \n",
       "\n",
       "                         TARSL2      DIS3  CTD.2256P15.2    ZNF283    FAHD2B  \\\n",
       "barcode                                                                        \n",
       "AAACCTGCAATAGCAA-1_06  0.660869 -0.059796       0.012625  0.007151  0.014972   \n",
       "AAACCTGCACAACTGT-1_06  0.248703  0.093876       0.005728  0.025548  0.009291   \n",
       "AAACCTGCAGCCTGTG-1_06 -0.085906  0.005892      -0.018601  0.030213 -0.042815   \n",
       "AAACCTGCAGTCAGCC-1_06  0.099969  0.084648       0.014874 -0.017749  0.041447   \n",
       "AAACCTGGTTCCTCCA-1_06 -0.028001  0.097279       0.011172  0.013083  0.161455   \n",
       "\n",
       "                       RP11.485G4.2  \n",
       "barcode                              \n",
       "AAACCTGCAATAGCAA-1_06      0.026571  \n",
       "AAACCTGCACAACTGT-1_06     -0.002509  \n",
       "AAACCTGCAGCCTGTG-1_06     -0.001975  \n",
       "AAACCTGCAGTCAGCC-1_06      0.007477  \n",
       "AAACCTGGTTCCTCCA-1_06     -0.006934  \n",
       "\n",
       "[5 rows x 5000 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genreate original BCR embeddings\n",
    "\n",
    "First, we download the pre-trained BCR encoder\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching 8 files: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 8/8 [00:28<00:00,  3.57s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Download Finished. Path /mnt/d/CoMBCR/CoMBCR/BCRencoder\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from CoMBCR.utils import download_BCRencoder\n",
    "\n",
    "download_BCRencoder()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Please clone or download the \"runberta.py\" in this github. This file is used to measure the original distances between BCRs. We recommend using our default pre-trained encoder, though any encoder can be used to encode BCRs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we directly used the original BCR embeddings under the exampledata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1014</th>\n",
       "      <th>1015</th>\n",
       "      <th>1016</th>\n",
       "      <th>1017</th>\n",
       "      <th>1018</th>\n",
       "      <th>1019</th>\n",
       "      <th>1020</th>\n",
       "      <th>1021</th>\n",
       "      <th>1022</th>\n",
       "      <th>1023</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAATAGCAA-1_06</th>\n",
       "      <td>-0.324237</td>\n",
       "      <td>-0.228240</td>\n",
       "      <td>-0.241070</td>\n",
       "      <td>0.951063</td>\n",
       "      <td>-0.688552</td>\n",
       "      <td>-0.024392</td>\n",
       "      <td>-0.455126</td>\n",
       "      <td>-0.142796</td>\n",
       "      <td>0.071060</td>\n",
       "      <td>0.872651</td>\n",
       "      <td>...</td>\n",
       "      <td>0.433799</td>\n",
       "      <td>-0.089641</td>\n",
       "      <td>0.199088</td>\n",
       "      <td>-0.619232</td>\n",
       "      <td>0.433234</td>\n",
       "      <td>-0.507768</td>\n",
       "      <td>0.094147</td>\n",
       "      <td>0.622088</td>\n",
       "      <td>-0.438514</td>\n",
       "      <td>-0.548302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACAACTGT-1_06</th>\n",
       "      <td>-0.836004</td>\n",
       "      <td>-0.281383</td>\n",
       "      <td>-1.062309</td>\n",
       "      <td>0.253088</td>\n",
       "      <td>-0.613961</td>\n",
       "      <td>0.173742</td>\n",
       "      <td>-0.924163</td>\n",
       "      <td>0.295760</td>\n",
       "      <td>0.215318</td>\n",
       "      <td>0.514034</td>\n",
       "      <td>...</td>\n",
       "      <td>0.113509</td>\n",
       "      <td>0.117395</td>\n",
       "      <td>-0.115959</td>\n",
       "      <td>-0.211876</td>\n",
       "      <td>0.793687</td>\n",
       "      <td>-0.666870</td>\n",
       "      <td>-0.068236</td>\n",
       "      <td>0.275671</td>\n",
       "      <td>0.094568</td>\n",
       "      <td>-0.396323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCCTGTG-1_06</th>\n",
       "      <td>-0.388252</td>\n",
       "      <td>-0.081444</td>\n",
       "      <td>-0.394865</td>\n",
       "      <td>1.480731</td>\n",
       "      <td>-0.288159</td>\n",
       "      <td>-0.132324</td>\n",
       "      <td>-0.263037</td>\n",
       "      <td>-0.209172</td>\n",
       "      <td>0.423102</td>\n",
       "      <td>0.569923</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.231537</td>\n",
       "      <td>-0.353619</td>\n",
       "      <td>0.118815</td>\n",
       "      <td>-1.144104</td>\n",
       "      <td>0.398278</td>\n",
       "      <td>0.067934</td>\n",
       "      <td>-0.181059</td>\n",
       "      <td>0.760884</td>\n",
       "      <td>-0.498049</td>\n",
       "      <td>-0.671089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGTCAGCC-1_06</th>\n",
       "      <td>-0.912942</td>\n",
       "      <td>0.000863</td>\n",
       "      <td>0.077432</td>\n",
       "      <td>0.949241</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.136745</td>\n",
       "      <td>-0.283928</td>\n",
       "      <td>-0.212797</td>\n",
       "      <td>0.105468</td>\n",
       "      <td>0.711840</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.587940</td>\n",
       "      <td>0.107624</td>\n",
       "      <td>-0.320966</td>\n",
       "      <td>-1.219826</td>\n",
       "      <td>0.783442</td>\n",
       "      <td>0.099214</td>\n",
       "      <td>0.109777</td>\n",
       "      <td>0.096048</td>\n",
       "      <td>-0.420305</td>\n",
       "      <td>-1.142159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTCCTCCA-1_06</th>\n",
       "      <td>-0.079208</td>\n",
       "      <td>-0.455230</td>\n",
       "      <td>-0.978580</td>\n",
       "      <td>1.132275</td>\n",
       "      <td>-0.574163</td>\n",
       "      <td>0.258197</td>\n",
       "      <td>-0.157592</td>\n",
       "      <td>-0.350995</td>\n",
       "      <td>0.127573</td>\n",
       "      <td>0.343525</td>\n",
       "      <td>...</td>\n",
       "      <td>0.351599</td>\n",
       "      <td>-0.057017</td>\n",
       "      <td>0.099804</td>\n",
       "      <td>-1.488126</td>\n",
       "      <td>0.658145</td>\n",
       "      <td>0.039580</td>\n",
       "      <td>-0.006186</td>\n",
       "      <td>-0.196100</td>\n",
       "      <td>-0.564170</td>\n",
       "      <td>-0.376476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1024 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              0         1         2         3         4  \\\n",
       "barcode                                                                   \n",
       "AAACCTGCAATAGCAA-1_06 -0.324237 -0.228240 -0.241070  0.951063 -0.688552   \n",
       "AAACCTGCACAACTGT-1_06 -0.836004 -0.281383 -1.062309  0.253088 -0.613961   \n",
       "AAACCTGCAGCCTGTG-1_06 -0.388252 -0.081444 -0.394865  1.480731 -0.288159   \n",
       "AAACCTGCAGTCAGCC-1_06 -0.912942  0.000863  0.077432  0.949241  0.020408   \n",
       "AAACCTGGTTCCTCCA-1_06 -0.079208 -0.455230 -0.978580  1.132275 -0.574163   \n",
       "\n",
       "                              5         6         7         8         9  ...  \\\n",
       "barcode                                                                  ...   \n",
       "AAACCTGCAATAGCAA-1_06 -0.024392 -0.455126 -0.142796  0.071060  0.872651  ...   \n",
       "AAACCTGCACAACTGT-1_06  0.173742 -0.924163  0.295760  0.215318  0.514034  ...   \n",
       "AAACCTGCAGCCTGTG-1_06 -0.132324 -0.263037 -0.209172  0.423102  0.569923  ...   \n",
       "AAACCTGCAGTCAGCC-1_06  0.136745 -0.283928 -0.212797  0.105468  0.711840  ...   \n",
       "AAACCTGGTTCCTCCA-1_06  0.258197 -0.157592 -0.350995  0.127573  0.343525  ...   \n",
       "\n",
       "                           1014      1015      1016      1017      1018  \\\n",
       "barcode                                                                   \n",
       "AAACCTGCAATAGCAA-1_06  0.433799 -0.089641  0.199088 -0.619232  0.433234   \n",
       "AAACCTGCACAACTGT-1_06  0.113509  0.117395 -0.115959 -0.211876  0.793687   \n",
       "AAACCTGCAGCCTGTG-1_06 -0.231537 -0.353619  0.118815 -1.144104  0.398278   \n",
       "AAACCTGCAGTCAGCC-1_06 -0.587940  0.107624 -0.320966 -1.219826  0.783442   \n",
       "AAACCTGGTTCCTCCA-1_06  0.351599 -0.057017  0.099804 -1.488126  0.658145   \n",
       "\n",
       "                           1019      1020      1021      1022      1023  \n",
       "barcode                                                                  \n",
       "AAACCTGCAATAGCAA-1_06 -0.507768  0.094147  0.622088 -0.438514 -0.548302  \n",
       "AAACCTGCACAACTGT-1_06 -0.666870 -0.068236  0.275671  0.094568 -0.396323  \n",
       "AAACCTGCAGCCTGTG-1_06  0.067934 -0.181059  0.760884 -0.498049 -0.671089  \n",
       "AAACCTGCAGTCAGCC-1_06  0.099214  0.109777  0.096048 -0.420305 -1.142159  \n",
       "AAACCTGGTTCCTCCA-1_06  0.039580 -0.006186 -0.196100 -0.564170 -0.376476  \n",
       "\n",
       "[5 rows x 1024 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcrori = pd.read_csv(\"exampledata/example_bcrori.csv\", index_col=\"barcode\")\n",
    "bcrori.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(bcr.index.tolist() == bcrori.index.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run CoMBCR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## steps\n",
    "CoMBCR contains the parameters as follows:\n",
    "1. bcrpath: (Required) The path to the BCR sequences file.\n",
    "2. rnapath:  (Required) The path to the gene expression file.\n",
    "3. bcroriginal: (Required) The path to bcr original embedding file.\n",
    "4. outdir:  (Required) The directory where the best checkpoint file and the output embeddings will be stored.\n",
    "5. checkpoint: default is \"best_network.pth\". This parameter specifies the name of the file where the best model checkpoint will be saved.\n",
    "6. lr: default is 1e-6.\n",
    "7. lam: default is 1e-1, the inner parameter (Parameter alpha in paper).\n",
    "8. batch_size: default 256.\n",
    "9. epochs: default 200.\n",
    "10. patience: default 15, the patience for early stopping.\n",
    "11. lr_step: default [40,100], These are the milestones for the MultiStepLR setting, which adjusts the learning rate at specified epochs.\n",
    "12. encoderprofile_in_dim: default 5000. Adjust this parameter if the number of input genes differs from 5000.\n",
    "13. separatebatch: The default is False. If set to True, BCRs from different samples will be treated as distinct BCRs. Ensure that your BCR input file contains a \"sample\" column if you choose to enable this option."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The codes below return numpy arrays for BCR embeddings and gex embeddings. Meanwhile, it will output \"bcrembedding.csv\" and \"gexemedding.csv\" under the outdir you designated.  \n",
    "\n",
    "If CUDA raised error, this is due to a crush with the previous loaded model. Please restart the jupyter notebook and directly run the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "learning rate is  1e-06\n",
      "Adjusting learning rate of group 0 to 1.0000e-06.\n",
      "Adjusting learning rate of group 0 to 1.0000e-06.\n",
      "Epoch:[0/1]\tloss:7.12308\tloss_cmc:3.473290\tloss_p2p:3.732409\tloss_b2b:0.003556\n"
     ]
    }
   ],
   "source": [
    "from CoMBCR.CoMBCR import CoMBCR_main\n",
    "\n",
    "bcremb, gexemb = CoMBCR_main(bcrpath=\"exampledata/example_bcr.csv\", \n",
    "            rnapath=\"exampledata/example_rna.csv\", \n",
    "            bcroriginal=\"exampledata/example_bcrori.csv\", \n",
    "            outdir=\"example_outdir\",\n",
    "            epochs=1,\n",
    "            batch_size=32,\n",
    "            encoderprofile_in_dim=5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1000, 256)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcremb.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1000, 256)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gexemb.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the output files \"bcrembedding.csv\" and \"gexembedding.csv\" located in the designated output directory. Please note that these CSV files directly store the numpy arrays and, as such, do not include any \"barcode\" column. When reading these files, ensure that you do not specify any index column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcremb = pd.read_csv(\"example_outdir/Embeddings/bcrembeddings\")\n",
    "gexemb = pd.read_csv(\"example_outdir/Embeddings/gexembeddings\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>246</th>\n",
       "      <th>247</th>\n",
       "      <th>248</th>\n",
       "      <th>249</th>\n",
       "      <th>250</th>\n",
       "      <th>251</th>\n",
       "      <th>252</th>\n",
       "      <th>253</th>\n",
       "      <th>254</th>\n",
       "      <th>255</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.015348</td>\n",
       "      <td>-0.087860</td>\n",
       "      <td>-0.013923</td>\n",
       "      <td>0.056350</td>\n",
       "      <td>-0.017100</td>\n",
       "      <td>0.068572</td>\n",
       "      <td>0.005855</td>\n",
       "      <td>-0.050495</td>\n",
       "      <td>-0.005957</td>\n",
       "      <td>0.008949</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.136085</td>\n",
       "      <td>-0.076334</td>\n",
       "      <td>-0.083860</td>\n",
       "      <td>-0.018186</td>\n",
       "      <td>-0.109076</td>\n",
       "      <td>-0.055655</td>\n",
       "      <td>0.039805</td>\n",
       "      <td>-0.069356</td>\n",
       "      <td>-0.029027</td>\n",
       "      <td>-0.031129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.124641</td>\n",
       "      <td>0.031098</td>\n",
       "      <td>0.027163</td>\n",
       "      <td>0.138613</td>\n",
       "      <td>-0.001733</td>\n",
       "      <td>0.038861</td>\n",
       "      <td>0.051714</td>\n",
       "      <td>0.011094</td>\n",
       "      <td>-0.049876</td>\n",
       "      <td>0.009269</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.106010</td>\n",
       "      <td>0.002925</td>\n",
       "      <td>-0.024259</td>\n",
       "      <td>-0.017597</td>\n",
       "      <td>-0.042338</td>\n",
       "      <td>-0.005912</td>\n",
       "      <td>0.011963</td>\n",
       "      <td>-0.059456</td>\n",
       "      <td>0.017165</td>\n",
       "      <td>0.023257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.035745</td>\n",
       "      <td>-0.049108</td>\n",
       "      <td>-0.037657</td>\n",
       "      <td>0.123449</td>\n",
       "      <td>0.000613</td>\n",
       "      <td>0.055640</td>\n",
       "      <td>0.086152</td>\n",
       "      <td>0.093559</td>\n",
       "      <td>0.029612</td>\n",
       "      <td>-0.020621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.182125</td>\n",
       "      <td>0.023510</td>\n",
       "      <td>-0.041843</td>\n",
       "      <td>-0.049026</td>\n",
       "      <td>-0.042913</td>\n",
       "      <td>0.086104</td>\n",
       "      <td>-0.060631</td>\n",
       "      <td>-0.022182</td>\n",
       "      <td>0.026433</td>\n",
       "      <td>0.007512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.087479</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>-0.032866</td>\n",
       "      <td>0.033859</td>\n",
       "      <td>0.027408</td>\n",
       "      <td>-0.012790</td>\n",
       "      <td>0.118991</td>\n",
       "      <td>0.037054</td>\n",
       "      <td>-0.002722</td>\n",
       "      <td>-0.013472</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.134776</td>\n",
       "      <td>0.024128</td>\n",
       "      <td>0.035573</td>\n",
       "      <td>-0.098892</td>\n",
       "      <td>0.000892</td>\n",
       "      <td>0.075088</td>\n",
       "      <td>0.008346</td>\n",
       "      <td>-0.024887</td>\n",
       "      <td>-0.018854</td>\n",
       "      <td>0.046138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.013984</td>\n",
       "      <td>-0.057683</td>\n",
       "      <td>-0.067793</td>\n",
       "      <td>0.080526</td>\n",
       "      <td>-0.009838</td>\n",
       "      <td>0.049828</td>\n",
       "      <td>0.068106</td>\n",
       "      <td>0.017483</td>\n",
       "      <td>0.001821</td>\n",
       "      <td>0.027285</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.127777</td>\n",
       "      <td>-0.004428</td>\n",
       "      <td>-0.013432</td>\n",
       "      <td>-0.056396</td>\n",
       "      <td>-0.009095</td>\n",
       "      <td>0.052927</td>\n",
       "      <td>-0.116035</td>\n",
       "      <td>0.022259</td>\n",
       "      <td>-0.001118</td>\n",
       "      <td>-0.014250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 256 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6  \\\n",
       "0  0.015348 -0.087860 -0.013923  0.056350 -0.017100  0.068572  0.005855   \n",
       "1  0.124641  0.031098  0.027163  0.138613 -0.001733  0.038861  0.051714   \n",
       "2  0.035745 -0.049108 -0.037657  0.123449  0.000613  0.055640  0.086152   \n",
       "3  0.087479  0.025641 -0.032866  0.033859  0.027408 -0.012790  0.118991   \n",
       "4  0.013984 -0.057683 -0.067793  0.080526 -0.009838  0.049828  0.068106   \n",
       "\n",
       "          7         8         9  ...       246       247       248       249  \\\n",
       "0 -0.050495 -0.005957  0.008949  ... -0.136085 -0.076334 -0.083860 -0.018186   \n",
       "1  0.011094 -0.049876  0.009269  ... -0.106010  0.002925 -0.024259 -0.017597   \n",
       "2  0.093559  0.029612 -0.020621  ... -0.182125  0.023510 -0.041843 -0.049026   \n",
       "3  0.037054 -0.002722 -0.013472  ... -0.134776  0.024128  0.035573 -0.098892   \n",
       "4  0.017483  0.001821  0.027285  ... -0.127777 -0.004428 -0.013432 -0.056396   \n",
       "\n",
       "        250       251       252       253       254       255  \n",
       "0 -0.109076 -0.055655  0.039805 -0.069356 -0.029027 -0.031129  \n",
       "1 -0.042338 -0.005912  0.011963 -0.059456  0.017165  0.023257  \n",
       "2 -0.042913  0.086104 -0.060631 -0.022182  0.026433  0.007512  \n",
       "3  0.000892  0.075088  0.008346 -0.024887 -0.018854  0.046138  \n",
       "4 -0.009095  0.052927 -0.116035  0.022259 -0.001118 -0.014250  \n",
       "\n",
       "[5 rows x 256 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcremb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>246</th>\n",
       "      <th>247</th>\n",
       "      <th>248</th>\n",
       "      <th>249</th>\n",
       "      <th>250</th>\n",
       "      <th>251</th>\n",
       "      <th>252</th>\n",
       "      <th>253</th>\n",
       "      <th>254</th>\n",
       "      <th>255</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.048509</td>\n",
       "      <td>-0.050780</td>\n",
       "      <td>0.012092</td>\n",
       "      <td>0.054801</td>\n",
       "      <td>-0.077096</td>\n",
       "      <td>-0.071795</td>\n",
       "      <td>0.120492</td>\n",
       "      <td>0.048243</td>\n",
       "      <td>-0.091755</td>\n",
       "      <td>-0.037499</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030512</td>\n",
       "      <td>-0.069922</td>\n",
       "      <td>-0.017588</td>\n",
       "      <td>-0.057541</td>\n",
       "      <td>-0.033826</td>\n",
       "      <td>-0.059658</td>\n",
       "      <td>-0.007020</td>\n",
       "      <td>-0.018299</td>\n",
       "      <td>0.034806</td>\n",
       "      <td>0.077380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.008329</td>\n",
       "      <td>-0.065424</td>\n",
       "      <td>0.061591</td>\n",
       "      <td>-0.017872</td>\n",
       "      <td>-0.059715</td>\n",
       "      <td>-0.040780</td>\n",
       "      <td>0.160052</td>\n",
       "      <td>-0.033229</td>\n",
       "      <td>-0.061964</td>\n",
       "      <td>-0.026953</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.013495</td>\n",
       "      <td>-0.101155</td>\n",
       "      <td>0.022401</td>\n",
       "      <td>-0.036999</td>\n",
       "      <td>-0.039105</td>\n",
       "      <td>-0.100446</td>\n",
       "      <td>-0.006467</td>\n",
       "      <td>-0.010011</td>\n",
       "      <td>0.036884</td>\n",
       "      <td>0.089870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.035611</td>\n",
       "      <td>-0.070885</td>\n",
       "      <td>0.032202</td>\n",
       "      <td>0.049993</td>\n",
       "      <td>-0.043414</td>\n",
       "      <td>-0.038092</td>\n",
       "      <td>0.121762</td>\n",
       "      <td>0.008126</td>\n",
       "      <td>-0.090482</td>\n",
       "      <td>-0.002106</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000275</td>\n",
       "      <td>-0.075901</td>\n",
       "      <td>0.017932</td>\n",
       "      <td>-0.014056</td>\n",
       "      <td>-0.033765</td>\n",
       "      <td>-0.051641</td>\n",
       "      <td>-0.003662</td>\n",
       "      <td>-0.027082</td>\n",
       "      <td>0.036087</td>\n",
       "      <td>0.060277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.063172</td>\n",
       "      <td>-0.066526</td>\n",
       "      <td>0.050900</td>\n",
       "      <td>0.066959</td>\n",
       "      <td>-0.041829</td>\n",
       "      <td>-0.060583</td>\n",
       "      <td>0.122334</td>\n",
       "      <td>-0.033201</td>\n",
       "      <td>-0.063927</td>\n",
       "      <td>0.005277</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.014794</td>\n",
       "      <td>-0.117377</td>\n",
       "      <td>0.009713</td>\n",
       "      <td>-0.037630</td>\n",
       "      <td>-0.051727</td>\n",
       "      <td>-0.059814</td>\n",
       "      <td>-0.001520</td>\n",
       "      <td>-0.034314</td>\n",
       "      <td>0.050281</td>\n",
       "      <td>0.058940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.012943</td>\n",
       "      <td>-0.077301</td>\n",
       "      <td>-0.010864</td>\n",
       "      <td>0.059000</td>\n",
       "      <td>-0.085983</td>\n",
       "      <td>-0.026742</td>\n",
       "      <td>0.133180</td>\n",
       "      <td>0.004001</td>\n",
       "      <td>-0.066243</td>\n",
       "      <td>-0.045583</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.045505</td>\n",
       "      <td>-0.119614</td>\n",
       "      <td>0.013246</td>\n",
       "      <td>-0.047752</td>\n",
       "      <td>-0.039822</td>\n",
       "      <td>-0.060070</td>\n",
       "      <td>0.023413</td>\n",
       "      <td>0.028433</td>\n",
       "      <td>0.026463</td>\n",
       "      <td>0.055106</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 256 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6  \\\n",
       "0 -0.048509 -0.050780  0.012092  0.054801 -0.077096 -0.071795  0.120492   \n",
       "1  0.008329 -0.065424  0.061591 -0.017872 -0.059715 -0.040780  0.160052   \n",
       "2 -0.035611 -0.070885  0.032202  0.049993 -0.043414 -0.038092  0.121762   \n",
       "3 -0.063172 -0.066526  0.050900  0.066959 -0.041829 -0.060583  0.122334   \n",
       "4 -0.012943 -0.077301 -0.010864  0.059000 -0.085983 -0.026742  0.133180   \n",
       "\n",
       "          7         8         9  ...       246       247       248       249  \\\n",
       "0  0.048243 -0.091755 -0.037499  ...  0.030512 -0.069922 -0.017588 -0.057541   \n",
       "1 -0.033229 -0.061964 -0.026953  ... -0.013495 -0.101155  0.022401 -0.036999   \n",
       "2  0.008126 -0.090482 -0.002106  ...  0.000275 -0.075901  0.017932 -0.014056   \n",
       "3 -0.033201 -0.063927  0.005277  ... -0.014794 -0.117377  0.009713 -0.037630   \n",
       "4  0.004001 -0.066243 -0.045583  ... -0.045505 -0.119614  0.013246 -0.047752   \n",
       "\n",
       "        250       251       252       253       254       255  \n",
       "0 -0.033826 -0.059658 -0.007020 -0.018299  0.034806  0.077380  \n",
       "1 -0.039105 -0.100446 -0.006467 -0.010011  0.036884  0.089870  \n",
       "2 -0.033765 -0.051641 -0.003662 -0.027082  0.036087  0.060277  \n",
       "3 -0.051727 -0.059814 -0.001520 -0.034314  0.050281  0.058940  \n",
       "4 -0.039822 -0.060070  0.023413  0.028433  0.026463  0.055106  \n",
       "\n",
       "[5 rows x 256 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gexemb.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
